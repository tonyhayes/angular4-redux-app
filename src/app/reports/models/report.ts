export interface ScheduleInfo {
  name: string; /*Lists your schedules by the name you assign to them*/
  repeats: RepeatTime; /*Describes how often the schedule is set to run.*/
  sourceFile: string; /*file associated with the schedule*/
  outputLocation: string; /* location that the scheduled report is saved*/
  lastRun: string; /* last time and date when the schedule was run*/
  nextRun: string; /* next time and date when the schedule will run again*/
  status: string; /*current Status of the schedule. The state can be either Normal or Paused.*/
}
export interface ReportInfo {
  name: string; /*Lists your report by the name you assign to them*/
  sourceFile: string; /*file associated with the report*/
  outputLocation: string; /* location that the report is saved*/
  lastRun: string; /* last time and date when the report was run*/
  status: string; /*current Status of the report. The state can be either Normal or Paused.*/
}

export interface RepeatTime {
  repeatTimeStart: string; /* all schedules are blocked from running from date*/
  repeatTimeEnds: string; /* all schedules are blocked from running to date*/
  repeats: string; /*Describes how often backout is active.*/
  repeatsEndBy: string; /*Describes when the  repeater ends.*/
}

export interface Report {
  id: string;
  blackoutTimesList: RepeatTime[];
  reportInfo: ReportInfo;
  scheduleInfo: ScheduleInfo;
}

export function generateMockReport(): Report {
  return {
    id: '1',
    blackoutTimesList: [
      {
        repeatTimeStart: '1513616909621' ,
        repeatTimeEnds: '1513616909621',
        repeats: 'Daily',
        repeatsEndBy: 'Never'
      }
    ],
    reportInfo: {
      name: 'report name',
      sourceFile: '/path/to/report',
      outputLocation: '/output/path/to/report',
      lastRun: '1513616909621',
      status: 'Normal',
    },
    scheduleInfo: {
      name: 'report schedule name',
      repeats: {
        repeatTimeStart: '1513616909621',
        repeatTimeEnds: '1513616909621',
        repeats: 'Daily',
        repeatsEndBy: 'Never'
      },
      sourceFile: '/path/to/report',
      outputLocation: '/output/path/to/report',
      lastRun: '1513616909621',
      nextRun: '1513616909621',
      status: 'Normal',
    },
  };
}
