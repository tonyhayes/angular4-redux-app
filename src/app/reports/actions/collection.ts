import { Action } from '@ngrx/store';
import { Report } from '../models/report';

export const ADD_REPORT = '[Collection] Add Report';
export const ADD_REPORT_SUCCESS = '[Collection] Add Report Success';
export const ADD_REPORT_FAIL = '[Collection] Add Report Fail';
export const REMOVE_REPORT = '[Collection] Remove Report';
export const REMOVE_REPORT_SUCCESS = '[Collection] Remove Report Success';
export const REMOVE_REPORT_FAIL = '[Collection] Remove Report Fail';
export const LOAD = '[Collection] Load';
export const LOAD_SUCCESS = '[Collection] Load Success';
export const LOAD_FAIL = '[Collection] Load Fail';

/**
 * Add Report to Collection Actions
 */
export class AddReport implements Action {
  readonly type = ADD_REPORT;

  constructor(public payload: Report) {}
}

export class AddReportSuccess implements Action {
  readonly type = ADD_REPORT_SUCCESS;

  constructor(public payload: Report) {}
}

export class AddReportFail implements Action {
  readonly type = ADD_REPORT_FAIL;

  constructor(public payload: Report) {}
}

/**
 * Remove Report from Collection Actions
 */
export class RemoveReport implements Action {
  readonly type = REMOVE_REPORT;

  constructor(public payload: Report) {}
}

export class RemoveReportSuccess implements Action {
  readonly type = REMOVE_REPORT_SUCCESS;

  constructor(public payload: Report) {}
}

export class RemoveReportFail implements Action {
  readonly type = REMOVE_REPORT_FAIL;

  constructor(public payload: Report) {}
}

/**
 * Load Collection Actions
 */
export class Load implements Action {
  readonly type = LOAD;
}

export class LoadSuccess implements Action {
  readonly type = LOAD_SUCCESS;

  constructor(public payload: Report[]) {}
}

export class LoadFail implements Action {
  readonly type = LOAD_FAIL;

  constructor(public payload: any) {}
}

export type Actions =
  | AddReport
  | AddReportSuccess
  | AddReportFail
  | RemoveReport
  | RemoveReportSuccess
  | RemoveReportFail
  | Load
  | LoadSuccess
  | LoadFail;
