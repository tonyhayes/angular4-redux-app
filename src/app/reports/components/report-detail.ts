import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Report } from '../models/report';

@Component({
  selector: 'analytics-report-detail',
  template: `
    <mat-card *ngIf="report">
      <mat-card-title-group>
        <mat-card-title>{{ title }}</mat-card-title>
      </mat-card-title-group>
        <mat-card-content>
          <p>Report Path : {{ sourceFile | bcEllipsis:40 }}</p>
          <p>Output Path : {{ outputLocation | bcEllipsis }}</p>
        </mat-card-content>
        <mat-card-content *ngIf="inCollection">
          <h5 mat-subheader>Schedule Info:</h5>
          <p>Repeats : {{ repeats }}</p>
          <p>End By : {{ repeatsEndBy }}</p>
          <p>Start Time : {{ repeatTimeStart | date:'medium' }}</p>
          <p>End Time : {{ repeatTimeEnds | date:'medium' }}</p>
        </mat-card-content>
      <mat-card-footer class="footer">
        <analytics-report-info [report]="report" [scheduled]="inCollection"></analytics-report-info>
      </mat-card-footer>
      <mat-card-actions align="start">
        <button mat-raised-button color="warn" *ngIf="inCollection" (click)="remove.emit(report)">
        Remove Report from Scheduler
        </button>

        <button mat-raised-button color="primary" *ngIf="!inCollection" (click)="add.emit(report)">
        Add Report to Scheduler
        </button>
      </mat-card-actions>
    </mat-card>

  `,
  styles: [
    `
    :host {
      display: flex;
      justify-content: center;
      margin: 75px 0;
    }
    mat-card {
      width: 500px;
      margin: 15px;
      margin-top: -35px;
      display: flex;
      flex-flow: column;
      justify-content: space-between;
    }

    @media only screen and (max-width: 768px) {
      mat-card {
        margin: 15px 0 !important;
      }
    }
    mat-card:hover {
      box-shadow: 3px 3px 16px -2px rgba(0, 0, 0, .5);
    }
    mat-card-title {
      margin-right: 10px;
    }
    mat-card-title-group {
      margin: 0;
    }
    a {
      color: inherit;
      text-decoration: none;
    }
    mat-card-content {
      margin-top: 15px;
      margin: 15px 0 0;
    }
    span {
      display: inline-block;
      font-size: 13px;
    }
    h5 {
      margin-bottom: 5px;
      font-size: 13px;
    }
    mat-card-actions {
      margin: 25px 0 0 !important;
    }
    mat-card-footer {
      padding: 0 25px 25px;
      position: relative;
    }
  `,
  ],
})
export class ReportDetailComponent {
  /**
   * Presentational components receive data through @Input() and communicate events
   * through @Output() but generally maintain no internal state of their
   * own. All decisions are delegated to 'container', or 'smart'
   * components before data updates flow back down.
   *
   * More on 'smart' and 'presentational' components: https://gist.github.com/btroncone/a6e4347326749f938510#utilizing-container-components
   */
  @Input() report: Report;
  @Input() inCollection: boolean;
  @Output() add = new EventEmitter<Report>();
  @Output() remove = new EventEmitter<Report>();

  /**
   * Tip: Utilize getters to keep templates clean
   */
  get id() {
    return this.report.id;
  }

  get title() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.name;
    }
    return 'Big Report I Run All The Time';
  }


  get description() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.name;
    }
    return 'Big Report I Run All The Time';
  }
  get repeats() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeats;
    }
    return 'Daily';
  }
  get repeatsEndBy() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeatsEndBy;
    }
    return 'Never';
  }
  get repeatTimeStart() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeatTimeStart;
    }
    return Date.now().toString();
  }
  get repeatTimeEnds() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.repeats.repeatTimeEnds;
    }
    return Date.now().toString();
  }
  get sourceFile() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.name;
    }
    return '/path/to/report';
  }
  get outputLocation() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.name;
    }
    return '/output/path/to/report';
  }}
