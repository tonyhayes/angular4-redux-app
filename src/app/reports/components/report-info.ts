import { Component, Input } from '@angular/core';

import { Report } from '../models/report';

@Component({
  selector: 'analytics-report-info',
  template: `
    <h5 mat-subheader>Report Info:</h5>
    <div *ngIf="scheduled">
      Next Run: {{ nextRun | date:'medium' }}
    </div>
    <div>
      Last Run: {{ lastRun | date:'medium' }}
    </div>
    <div>
      Status: {{ status }}
    </div>
  `,
  styles: [
    `
    h5 {
      margin-bottom: 5px;
    }
  `,
  ],
})
export class ReportInfoComponent {
  @Input() report: Report;
  @Input() scheduled: boolean = false;

  get nextRun() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.nextRun;
    }
    return Date.now();
  }
  get lastRun() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.lastRun;
    }
    return Date.now();
  }
  get status() {
    if (this.report.scheduleInfo) {
      return this.report.scheduleInfo.status;
    }
    return 'Normal';
  }
}
