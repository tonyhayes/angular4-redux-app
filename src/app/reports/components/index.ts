import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';

import { ReportInfoComponent } from './report-info';
import { ReportDetailComponent } from './report-detail';
import { ReportPreviewComponent } from './report-preview';
import { ReportPreviewListComponent } from './report-preview-list';
import { ReportSearchComponent } from './report-search';

import { PipesModule } from '../../shared/pipes';
import { MaterialModule } from '../../material';

export const COMPONENTS = [
  ReportInfoComponent,
  ReportDetailComponent,
  ReportPreviewComponent,
  ReportPreviewListComponent,
  ReportSearchComponent,
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    RouterModule,
    PipesModule,
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class ComponentsModule {}
