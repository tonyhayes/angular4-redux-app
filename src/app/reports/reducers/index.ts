import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromSearch from './search';
import * as fromReports from './reports';
import * as fromCollection from './collection';
import * as fromRoot from '../../reducers';

export interface ReportsState {
  search: fromSearch.State;
  reports: fromReports.State;
  collection: fromCollection.State;
}

export interface State extends fromRoot.State {
  'reports': ReportsState;
}

export const reducers = {
  search: fromSearch.reducer,
  reports: fromReports.reducer,
  collection: fromCollection.reducer,
};

/**
 * A selector function is a map function factory. We pass it parameters and it
 * returns a function that maps from the larger state tree into a smaller
 * piece of state. This selector simply selects the `reports` state.
 *
 * Selectors are used with the `select` operator.
 *
 * ```ts
 * class MyComponent {
 * 	constructor(state$: Observable<State>) {
 * 	  this.reportsState$ = state$.select(getReportsState);
 * 	}
 * }
 * ```
 */

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
*/
export const getReportsState = createFeatureSelector<ReportsState>('reports');

/**
 * Every reducer module exports selector functions, however child reducers
 * have no knowledge of the overall state tree. To make them usable, we
 * need to make new selectors that wrap them.
 *
 * The createSelector function creates very efficient selectors that are memoized and
 * only recompute when arguments change. The created selectors can also be composed
 * together to select different pieces of state.
 */
export const getReportEntitiesState = createSelector(
  getReportsState,
  state => state.reports
);

export const getSelectedReportId = createSelector(
  getReportEntitiesState,
  fromReports.getSelectedId
);

/**
 * Adapters created with @ngrx/entity generate
 * commonly used selector functions including
 * getting all ids in the record set, a dictionary
 * of the records by id, an array of records and
 * the total number of records. This reducers boilerplate
 * in selecting records from the entity state.
 */
export const {
  selectIds: getReportIds,
  selectEntities: getReportEntities,
  selectAll: getAllReports,
  selectTotal: getTotalReports,
} = fromReports.adapter.getSelectors(getReportEntitiesState);

export const getSelectedReport = createSelector(
  getReportEntities,
  getSelectedReportId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);

/**
 * Just like with the reports selectors, we also have to compose the search
 * reducer's and collection reducer's selectors.
 */
export const getSearchState = createSelector(
  getReportsState,
  (state: ReportsState) => state.search
);

export const getSearchReportIds = createSelector(
  getSearchState,
  fromSearch.getIds
);
export const getSearchQuery = createSelector(
  getSearchState,
  fromSearch.getQuery
);
export const getSearchLoading = createSelector(
  getSearchState,
  fromSearch.getLoading
);
export const getSearchError = createSelector(
  getSearchState,
  fromSearch.getError
);

/**
 * Some selector functions create joins across parts of state. This selector
 * composes the search result IDs to return an array of reports in the store.
 */
export const getSearchResults = createSelector(
  getReportEntities,
  getSearchReportIds,
  (reports, searchIds) => {
    return searchIds.map(id => reports[id]);
  }
);

export const getCollectionState = createSelector(
  getReportsState,
  (state: ReportsState) => state.collection
);

export const getCollectionLoaded = createSelector(
  getCollectionState,
  fromCollection.getLoaded
);
export const getCollectionLoading = createSelector(
  getCollectionState,
  fromCollection.getLoading
);
export const getCollectionReportIds = createSelector(
  getCollectionState,
  fromCollection.getIds
);

export const getReportCollection = createSelector(
  getReportEntities,
  getCollectionReportIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const isSelectedReportInCollection = createSelector(
  getCollectionReportIds,
  getSelectedReportId,
  (ids, selected) => {
    return ids.indexOf(selected) > -1;
  }
);
