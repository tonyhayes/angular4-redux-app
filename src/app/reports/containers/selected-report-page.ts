import { Component, ChangeDetectionStrategy } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';

import * as fromReports from '../reducers';
import * as collection from '../actions/collection';
import { Report } from '../models/report';

@Component({
  selector: 'analytics-selected-report-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <analytics-report-detail
      [report]="report$ | async"
      [inCollection]="isSelectedReportInCollection$ | async"
      (add)="addToCollection($event)"
      (remove)="removeFromCollection($event)">
    </analytics-report-detail>
  `,
})
export class SelectedReportPageComponent {
  report$: Observable<Report>;
  isSelectedReportInCollection$: Observable<boolean>;

  constructor(private store: Store<fromReports.State>) {
    this.report$ = store.select(fromReports.getSelectedReport);
    this.isSelectedReportInCollection$ = store.select(
      fromReports.isSelectedReportInCollection
    );
  }

  addToCollection(report: Report) {
    this.store.dispatch(new collection.AddReport(report));
  }

  removeFromCollection(report: Report) {
    this.store.dispatch(new collection.RemoveReport(report));
  }
}
