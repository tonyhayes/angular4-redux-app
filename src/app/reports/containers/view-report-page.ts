import 'rxjs/add/operator/map';
import 'rxjs/add/operator/pluck';
import { Component, OnDestroy, ChangeDetectionStrategy } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { Store } from '@ngrx/store';
import { Subscription } from 'rxjs/Subscription';

import * as fromReports from '../reducers';
import * as report from '../actions/report';

/**
 * Note: Container components are also reusable. Whether or not
 * a component is a presentation component or a container
 * component is an implementation detail.
 *
 * The View Report Page's responsibility is to map router params
 * to a 'Select' report action. Actually showing the selected
 * report remains a responsibility of the
 * SelectedReportPageComponent
 */
@Component({
  selector: 'analytics-view-report-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <analytics-selected-report-page></analytics-selected-report-page>
  `,
})
export class ViewReportPageComponent implements OnDestroy {
  actionsSubscription: Subscription;

  constructor(store: Store<fromReports.State>, route: ActivatedRoute) {
    this.actionsSubscription = route.params
      .map(params => new report.Select(params.id))
      .subscribe(store);
  }

  ngOnDestroy() {
    this.actionsSubscription.unsubscribe();
  }
}
