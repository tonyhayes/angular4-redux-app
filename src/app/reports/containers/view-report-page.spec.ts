import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { ActivatedRoute } from '@angular/router';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { MatCardModule } from '@angular/material';

import { ViewReportPageComponent } from './view-report-page';
import * as report from '../actions/report';
import * as fromReports from '../reducers';
import { SelectedReportPageComponent } from './selected-report-page';
import { ReportDetailComponent } from '../components/report-detail';
import { ReportInfoComponent } from '../components/report-info';
import { AddCommasPipe } from '../../shared/pipes/add-commas';
import { EllipsisPipe } from '../../shared/pipes/ellipsis';

describe('View Report Page', () => {
  let params = new BehaviorSubject({});
  let fixture: ComponentFixture<ViewReportPageComponent>;
  let store: Store<fromReports.State>;
  let instance: ViewReportPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [MatCardModule],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: { params },
        },
        {
          provide: Store,
          useValue: {
            select: jest.fn(),
            next: jest.fn(),
          },
        },
      ],
      declarations: [
        ViewReportPageComponent,
        SelectedReportPageComponent,
        ReportDetailComponent,
        ReportInfoComponent,
        AddCommasPipe,
        EllipsisPipe
      ],
    });

    fixture = TestBed.createComponent(ViewReportPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a report.Select action on init', () => {
    const action = new report.Select('2');
    params.next({ id: '2' });

    fixture.detectChanges();

    expect(store.next).toHaveBeenLastCalledWith(action);
  });
});
