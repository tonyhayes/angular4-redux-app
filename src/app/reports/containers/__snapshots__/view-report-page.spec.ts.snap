// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Report Page should compile 1`] = `
<analytics-view-report-page
  actionsSubscription={[Function Subscriber]}
>
  
      <analytics-selected-report-page>
    
      
    <analytics-report-detail
      _nghost-c0=""
    >
      
      
      
      
  
    
    </analytics-report-detail>
    
    
  </analytics-selected-report-page>
    
</analytics-view-report-page>
`;
